python/ray/dataframe/dataframe.py

```diff
@@ -20,22 +18,21 @@ def __init__(self, df, columns, index=None):
                 partitions.
             columns (pandas.Index): The column names for this dataframe, in
                 pandas Index object.
+            index (pandas.Index or list): The row index for this dataframe.
         """
         assert(len(df) > 0)
 
         self._df = df
-        # TODO: Clean up later.
-        # We will call get only when we access the object (and only once).
-        self._lengths = \
-            ray.get([_deploy_func.remote(_get_lengths, d) for d in self._df])
+        self._lengths = [_deploy_func.remote(_get_lengths, d)
+                         for d in self._df]
         self.columns = columns
 
-        if index is None:
-            self._index = self._default_index()
-        else:
-            self._index = index
+        # this _index object is a pd.DataFrame
+        # and we use that DataFrame's Index to index the rows.
+        self._index = self._default_index()
 
-        self._pd_index = None
+        if index is not None:
```

Comment:
This gives us automatic error checking from Pandas. We first create a default index, then reassign to verify that what was passed in is correct.